### PR TITLE
Wp import custom fields from csv

### PR DESCRIPTION
**Pull request by kiyote33**
image

Added 7th Tradition Distribution Information to our Code For Recovery demo, using the ACF plugin. The ACF tool allows for easy export/import of json to add custom fields to tsml.

Still testing against other plugins, but currently tests well with ACF.

### CODE REVIEW

File: 12-step-meeting-list.php [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 define('TSML_PATH', plugin_dir_path(__FILE__));
 
-define('TSML_VERSION', '3.14.35');
+define('TSML_VERSION', '3.14.33');
```
Comment (joshreisner) [inline 12-step-meeting-list.php:21]: dont think this change is intentional

File: includes/admin_import.php [hunk 2]
```diff
@@ -17,7 +17,7 @@ function tsml_import_page()
 
         //if posting a CSV, check for errors and add it to the import buffer
         if (isset($_FILES['tsml_import']) && isset($_POST['tsml_nonce']) && wp_verify_nonce($_POST['tsml_nonce'], $tsml_nonce)) {
-            ini_set('auto_detect_line_endings', 1); //to handle mac \r line endings
+            @ini_set('auto_detect_line_endings', 1); //to handle mac \r line endings
```
Comment (joshreisner) [inline includes/admin_import.php:20]: don't think we need this change

File: includes/admin_import.php [hunk 3]
```diff
@@ -92,7 +92,7 @@ function tsml_import_page()
 
                             //associate
                             $meeting = array_combine($header, $meeting);
-                        }
+                        }
```
Comment (joshreisner) [inline includes/admin_import.php:95]: don't need

File: includes/ajax.php [hunk 4]
```diff
@@ -478,9 +478,19 @@ function tsml_ajax_geocode()
...
+                add_post_meta($meeting_id, $key, $meeting[$key]);
+            }
+        }
+
+        //when found, update ACF custom meeting field ids
+        if (!empty($tsml_custom_meeting_fields)) {
+            $update_query = 'UPDATE ' . $wpdb->postmeta . ' ' .
+                'SET post_id = ' . $meeting_id . ' ' .
+                'WHERE post_id = ' . $meeting['id'];
+            $results = $wpdb->get_results($update_query);
```
Comment (joshreisner) [inline includes/ajax.php:493]: don't think we need these changes

File: includes/functions.php [hunk 5]
```diff
@@ -1191,11 +1191,11 @@ function tsml_get_meetings($arguments = [], $from_cache = true, $full_export = f
                 'author' => get_the_author_meta('user_login', $post->post_author)
             ], $locations[$post->post_parent]);
 
-            // include user-defined meeting fields
+            /* include user-defined meeting fields */
             if (!empty($tsml_custom_meeting_fields)) {
                 foreach ($tsml_custom_meeting_fields as $field => $title) {
                     if (!empty($meeting_meta[$post->ID][$field])) {
-                        $meeting[$field] = $meeting_meta[$post->ID][$field];
+                        $meeting[$field] = isset($meeting_meta[$post->ID][$field]) ? $meeting_meta[$post->ID][$field] : $meeting[$field];
```
Comment (joshreisner) [inline includes/functions.php:1198]: we just established that `!empty($meeting_meta[$post->ID][$field])` so this change should not be needed

File: includes/functions.php [hunk 6]
```diff
@@ -1445,9 +1445,9 @@ function tsml_meeting_types($types)
 //called from admin_import.php (both CSV and JSON)
 function tsml_import_buffer_set($meetings, $data_source_url = null, $data_source_parent_region_id = null)
 {
-    global $tsml_programs, $tsml_program, $tsml_days, $tsml_meeting_attendance_options, $tsml_data_sources;
+    global $tsml_programs, $tsml_program, $tsml_days, $tsml_meeting_attendance_options, $tsml_data_sources, $tsml_custom_meeting_fields;
 
-    if (strpos($data_source_url, "sheets.googleapis.com") !== false) {
+    if (@strpos($data_source_url, "sheets.googleapis.com") !== false) {
```
Comment (joshreisner) [inline includes/functions.php:1450]: don't think we need

**Review by joshreisner — CHANGES_REQUESTED** [review]

File: includes/ajax.php (hunk 4 above)
Comment (kiyote33) [reply includes/ajax.php:493]: This change is the key to making it work. Not able to achieve the import of custom fields with it!

**Comment by kiyote33** [comment]
Hey @josh, per your review, the only major issue I have with it is your thought 'of the update of the post_id to reset from the imported id to the new post_id' being unnecessary. It is required. Here's the code:         /* import custom user-defined meeting fields from CSV */
        if (!empty($tsml_custom_meeting_fields)) {
            foreach ($tsml_custom_meeting_fields as $field => $title) {
                if (array_key_exists($field, $meeting) && !empty($meeting[$field])) {
                    $import_id = (int) $meeting['id'];
                    $meta_value = (string) $meeting[$field];
                    add_post_meta($import_id, $field, $meta_value);
                }
            }
            /* update custom meeting field ids to new Post ID */
            $update_query = 'UPDATE ' . $wpdb->postmeta . ' ' .
                'SET post_id = ' . $meeting_id . ' ' .
                'WHERE post_id = ' . $meeting['id'];
            $results = $wpdb->get_results($update_query);
        }

File: includes/functions.php [hunk 7]
```diff
@@ -1684,7 +1684,18 @@ function tsml_import_buffer_set($meetings, $data_source_url = null, $data_source
...
+        if (!empty($tsml_custom_meeting_fields)) {
+            foreach ($tsml_custom_meeting_fields as $field => $title) {
+                if (array_key_exists($field, $meetings[$i]) && !empty($meetings[$i][$field])) {
+                    $meeting_id = (int) $meetings[$i]['id'];
+                    $meta_value = (string) $meetings[$i][$field];
+                    add_post_meta($meeting_id, $field, $meta_value);
+                }
+            }
+        }
+
```
[the remark this answers is not in the document]
Comment (kiyote33) [reply includes/functions.php:1698]: Done!

File: includes/save.php [hunk 8]
```diff
@@ -425,6 +425,17 @@
...
+        if (!$update || @strcmp($old_meeting->{$field}, $_POST[$field]) !== 0) {
+            $changes[] = $field;
+        }
+        if (empty($_POST[$field])) {
+            delete_post_meta($post->ID, $field);
+        } else {
+            update_post_meta($post->ID, $field, tsml_sanitize($title, $_POST[$field]));
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (kiyote33) [reply includes/save.php:438]: OK. Wasn't sure. Will remove it.

[the remark this answers is not in the document]
Comment (kiyote33) [reply includes/save.php:438]: Seems to work OK without it.

File: includes/ajax.php [hunk 9]
```diff
@@ -479,8 +479,26 @@ function tsml_ajax_geocode()
...
+           }
+       }
+
+        //import custom user-defined meeting fields from CSV 
+        if (!empty($tsml_custom_meeting_fields)) {
+            foreach ($tsml_custom_meeting_fields as $field => $title) {
+                if (array_key_exists($field, $meeting) && !empty($meeting[$field])) {
+                    $import_id = (int) $meeting['id'];
+                    $meta_value = (string) $meeting[$field];
+                    add_post_meta($import_id, $field, $meta_value);
```
Comment (joshreisner) [inline includes/ajax.php:494]: i guess i'm a little confused by the difference between $meeting['id'] and $meeting_id - the former comes from the buffer and the latter is what we just created? 

could you instead make this `add_post_meta($meeting_id, $field, $meta_value);`? that way you might not need the `$import_id` variable or the SQL query below

Comment (kiyote33) [reply includes/ajax.php:494]: That's what I originally had, but no joy. I spent numerous hours trying to figure a work around. The import_id is just for clarity.

File: 12-step-meeting-list.php (hunk 1 above)
Comment (joshreisner) [reply 12-step-meeting-list.php:21]: this doesn't seem to be resolved yet

File: includes/functions.php [hunk 10]
```diff
@@ -1445,7 +1445,7 @@ function tsml_meeting_types($types)
 //called from admin_import.php (both CSV and JSON)
 function tsml_import_buffer_set($meetings, $data_source_url = null, $data_source_parent_region_id = null)
 {
-    global $tsml_programs, $tsml_program, $tsml_days, $tsml_meeting_attendance_options, $tsml_data_sources;
+    global $tsml_programs, $tsml_program, $tsml_days, $tsml_meeting_attendance_options, $tsml_data_sources, $tsml_custom_meeting_fields;
```
Comment (joshreisner) [inline includes/functions.php:1448]: dont' think you need this import now that the code is moved

File: includes/ajax.php [hunk 11]
```diff
@@ -479,8 +479,24 @@ function tsml_ajax_geocode()
         }
 
         //add custom meeting fields if available
-        foreach (['types', 'data_source', 'conference_url', 'conference_url_notes', 'conference_phone', 'conference_phone_notes'] as $key) {
-            if (!empty($meeting[$key])) add_post_meta($meeting_id, $key, $meeting[$key]);
+       foreach (['types', 'data_source', 'conference_url', 'conference_url_notes', 'conference_phone', 'conference_phone_notes'] as $key) {
+           if (!empty($meeting[$key])) if (!empty($meeting[$key])) add_post_meta($meeting_id, $key, $meeting[$key]);
```
Comment (joshreisner) [inline includes/ajax.php:483]: looks like a second `if (!empty($meeting[$key]))` got added here

Comment (kiyote33) [reply includes/ajax.php:483]: Second IF statement removed.

File: 12-step-meeting-list.php (hunk 1 above)
Comment (kiyote33) [reply 12-step-meeting-list.php]: Fixed

File: includes/ajax.php (hunk 9 above)
Comment (kiyote33) [reply includes/ajax.php]: add_post_meta($meeting_id, $field, $meta_value);  coded and tested...

**Comment by kiyote33** [comment]
Resolved now
________________________________

**Comment by kiyote33** [comment]
Fixed
________________________________

**Review by joshreisner — APPROVED** [review]
💥

**Comment by kiyote33** [comment]
This closes issue #1416 which evolved from discussion #1413.

**Pull request merged**